src/basic/dirent-util.h

```diff
@@ -36,6 +36,7 @@ struct dirent *readdir_no_dot(DIR *dirp);
 /* Only if 64bit off_t is enabled struct dirent + struct dirent64 are actually the same. We require this, and
  * we want them to be interchangeable to make getdents64() work, hence verify that. */
 assert_cc(_FILE_OFFSET_BITS == 64);
+#if HAVE_STRUCT_DIRENT64
```

Comment:
this needs a comment. 

also, rather than remove the checks I think the right approach is to make musl differences go away, i.e. musl users should just compile their stuff with `-Ddirent64=dirent`...

Why does that have to leak in our codebase?

I am still not on board accepting musl's API definitions when they depart from glibc and also make little sense (i.e. musl's definition of getdents() is just weird, because it pretends the buffer was an array of dirent structs, but it isn't, it's a packed concatenation of variable size structs with suffixes.

If musl departs from Linux API (as defined by glibc), then they should carry he burden for that really, not us.

Reply:
I think adding comment is a good idea as it explains the context better for everyone.

@poettering the LFS  versions of functions were transitionary in glibc and musl came after this and decided to use 64bit off_t from day 1. I think musl is not diverging but its not implementing all of glibc-specific APIs, I think you will be open to such implementations which are portable across glibc and musl and we do not lose the functionality :)

Reply:
Yeah, even with glibc, dirent64 may be missing. (Though, with our default setting, it always exists.)
Hence, I approve the patch.

Reply:
we do not support builds without FILE_OFFSET_BITS=64, this is expressly encoded in our build system. on glibc dirent64 is then defined even on archs that never had to deal with 32bit in any way. I wish musl would just stop being so dickish and not even provide compat definitions there.

> @poettering the LFS versions of functions were transitionary in glibc and musl came after this and decided to use 64bit off_t from day 1.

Nope. There is no getdents() in glibc. There's only getdents64(). It's not transitionary for glibc at all in this regard.

Reply:
And iirc even musl calls the wrappe getdents64() though because it doesn#t play nice it's defined with a different prototype because they just don't care.

Reply:
> Really, the *64 functions are a crutch anywhere and ideally we'd just use the native versions instead.

This is a simplistic view of the world, because while for most such syscalls there are both versions around and supported by libcs, for the getdents64() library call in glibc there never was and still isn't a getdents() counterpart. You *have* to call the function with the "64" in the name or you simply can't call the function at all...

Reply:
I meant the dual ABI ones.